jellyfin/jellyfin-web
# Load gamepadtokey.js only when is necessary

This PR aims to fix all the problems and lags caused by the resource-hungry ``gamepadtokey.js`` script, only loading it when a gamepad is connected and detected.

This is a WIP because I don't have a controller at hand, so everything here has been done based in assumptions and hasn't been thoroughly tested. All the help that can be provided for testing this with a gamepad is more than appreciated

**Changes**

The logic that started the ``gamepadtokey.js`` in ``serverNotifications.js`` (L201) was always returning undefined, so the function ``enableNativeGamepadKeyMapping`` was always false. The second check done in ``isGamepadSupported`` was always true, as all modern browsers support gamepads, although that doesn't mean that a gamepad is connected. 

Now, these checks have been replaced in ``L216``, calling the ``getGamepads()`` method. If the browser supports that function, it will be true, else undefined, not triggering the gamepadtokey.js script. That way, two checks have been replaced by one.

In that if statement, we also check the first index of the array returned, as [described here](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad_API/Using_the_Gamepad_API#querying), so we know if a gamepad is connected or not.

Also, I removed all those checks in ``gamepadtokey.js``, as they're useless there because all the gamepad checks have been made in ``serverNotifications.js``. We also removed the function ``webkitGetGamepads()`` as it's completely deprecated.

**Concerns**

I have some doubts that I want to cover/discuss
* I also disabled ``mouseManager.js`` because it doesn't seem to do anything, at least that's what I found. Any clue on what it does?
* Any reason the proposed solution in #525 wasn't merged in the code? We can add that change to this PR as well.

**Issues**
Fixes #602 and should improve #545, as well as overall performance of the web interface

**To test**

* Connecting a gamepad and entering Jellyfin
* Connecting a gamepad while Jellyfin is running and check that is handled correctly and the gamepadtokey.js script is started.
* Disconnect a gamepad while Jellyfin is running and check that the gamepad script is killed.
